source/extensions/transport_sockets/tap/tap.cc

```diff
@@ -11,59 +11,82 @@ namespace Extensions {
+    : config_(config), transport_socket_(std::move(transport_socket)) {}
 
 void TapSocket::setTransportSocketCallbacks(Network::TransportSocketCallbacks& callbacks) {
-  callbacks_ = &callbacks;
   transport_socket_->setTransportSocketCallbacks(callbacks);
+  tapper_ = config_ ? config_->createPerSocketTapper(callbacks.connection()) : nullptr;
 }
 
 std::string TapSocket::protocol() const { return transport_socket_->protocol(); }
 
 bool TapSocket::canFlushClose() { return transport_socket_->canFlushClose(); }
 
 void TapSocket::closeSocket(Network::ConnectionEvent event) {
-  // The caller should have invoked setTransportSocketCallbacks() prior to this.
-  ASSERT(callbacks_ != nullptr);
-  auto* connection = trace_.mutable_connection();
-  connection->set_id(callbacks_->connection().id());
-  Network::Utility::addressToProtobufAddress(*callbacks_->connection().localAddress(),
-                                             *connection->mutable_local_address());
-  Network::Utility::addressToProtobufAddress(*callbacks_->connection().remoteAddress(),
-                                             *connection->mutable_remote_address());
-  const bool text_format =
-      format_ == envoy::config::transport_socket::tap::v2alpha::FileSink::PROTO_TEXT;
-  const std::string path = fmt::format("{}_{}.{}", path_prefix_, callbacks_->connection().id(),
-                                       text_format ? "pb_text" : "pb");
-  ENVOY_LOG_MISC(debug, "Writing socket trace for [C{}] to {}", callbacks_->connection().id(),
-                 path);
-  ENVOY_LOG_MISC(trace, "Socket trace for [C{}]: {}", callbacks_->connection().id(),
-                 trace_.DebugString());
-  std::ofstream proto_stream(path);
-  if (text_format) {
-    proto_stream << trace_.DebugString();
-  } else {
-    trace_.SerializeToOstream(&proto_stream);
+  if (tapper_ != nullptr) {
+    tapper_->closeSocket(event);
   }
+
   transport_socket_->closeSocket(event);
 }
 
 Network::IoResult TapSocket::doRead(Buffer::Instance& buffer) {
   Network::IoResult result = transport_socket_->doRead(buffer);
-  if (result.bytes_processed_ > 0) {
+  if (tapper_ != nullptr && result.bytes_processed_ > 0) {
     // TODO(htuch): avoid linearizing
-    char* data = static_cast<char*>(buffer.linearize(buffer.length())) +
-                 (buffer.length() - result.bytes_processed_);
-    auto* event = trace_.add_events();
-    event->mutable_timestamp()->MergeFrom(Protobuf::util::TimeUtil::NanosecondsToTimestamp(
-        std::chrono::duration_cast<std::chrono::nanoseconds>(
-            time_system_.systemTime().time_since_epoch())
-            .count()));
-    event->mutable_read()->set_data(data, result.bytes_processed_);
+    const char* data = static_cast<const char*>(buffer.linearize(buffer.length())) +
+                       (buffer.length() - result.bytes_processed_);
+    tapper_->onRead(absl::string_view(data, result.bytes_processed_));
```

Comment:
I thought about this, but then I can't linearize and I have to solve this problem in this PR, and I would prefer to not pass a non-const buffer and linearize in the tapper. I will add some more comments in the TODO on fixing this in a follow up?

Reply:
actually now we have [toString](https://github.com/envoyproxy/envoy/blob/master/include/envoy/buffer/buffer.h#L193) in `const Buffer::Instance` so I think you can just address that easily?

Reply:
Sure, can do that. It will involve an extra copy over what is in this PR, but given the perf issue already exists I can move the perf TODO into the tapper.